orders-accounting/src/tests/operations.rs

```diff
@@ -0,0 +1,558 @@
+    let mut rng = make_seedable_rng(seed);
+
+    let order_id = OrderId::random_using(&mut rng);
+    let order_data = make_order_data(&mut rng);
+
+    let mut storage = InMemoryOrdersAccounting::from_values(
+        BTreeMap::from_iter([(order_id, order_data.clone())]),
+        BTreeMap::from_iter([(order_id, output_value_amount(order_data.ask()).unwrap())]),
+        BTreeMap::from_iter([(order_id, output_value_amount(order_data.give()).unwrap())]),
+    );
+    let mut db = OrdersAccountingDB::new(&mut storage);
+    let mut cache = OrdersAccountingCache::new(&db);
+
+    // try to cancel non-existing order
+    {
+        let random_order = OrderId::random_using(&mut rng);
+        let result = cache.cancel_order(random_order);
+        assert_eq!(
+            result.unwrap_err(),
+            Error::AttemptedCancelNonexistingOrderData(random_order)
+        );
+    }
+
+    let _ = cache.cancel_order(order_id).unwrap();
+
+    db.batch_write_orders_data(cache.consume()).unwrap();
+
+    assert_eq!(InMemoryOrdersAccounting::new(), storage);
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn cancel_order_twice(#[case] seed: Seed) {
+    let mut rng = make_seedable_rng(seed);
+
+    let order_id = OrderId::random_using(&mut rng);
+    let order_data = make_order_data(&mut rng);
+
+    let storage = InMemoryOrdersAccounting::from_values(
+        BTreeMap::from_iter([(order_id, order_data.clone())]),
+        BTreeMap::from_iter([(order_id, output_value_amount(order_data.ask()).unwrap())]),
+        BTreeMap::from_iter([(order_id, output_value_amount(order_data.give()).unwrap())]),
+    );
+    let db = OrdersAccountingDB::new(&storage);
+    let mut cache = OrdersAccountingCache::new(&db);
+
+    let _ = cache.cancel_order(order_id).unwrap();
+
+    assert_eq!(
+        cache.cancel_order(order_id,),
+        Err(Error::AttemptedCancelNonexistingOrderData(order_id))
+    );
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn cancel_order_and_undo(#[case] seed: Seed) {
+    let mut rng = make_seedable_rng(seed);
+
+    let order_id = OrderId::random_using(&mut rng);
+    let order_data = make_order_data(&mut rng);
+
+    let mut storage = InMemoryOrdersAccounting::from_values(
+        BTreeMap::from_iter([(order_id, order_data.clone())]),
+        BTreeMap::from_iter([(order_id, output_value_amount(order_data.ask()).unwrap())]),
+        BTreeMap::from_iter([(order_id, output_value_amount(order_data.give()).unwrap())]),
+    );
+    let original_storage = storage.clone();
+    let mut db = OrdersAccountingDB::new(&mut storage);
+    let mut cache = OrdersAccountingCache::new(&db);
+
+    let undo = cache.cancel_order(order_id).unwrap();
+
+    assert_eq!(None, cache.get_order_data(&order_id).unwrap().as_ref());
+    assert_eq!(
+        Some(Amount::ZERO),
+        cache.get_ask_balance(&order_id).unwrap()
+    );
+    assert_eq!(
+        Some(Amount::ZERO),
+        cache.get_give_balance(&order_id).unwrap()
+    );
```

Comment:
This is probably a good example where canceling doesn't lead to None... all these accounting primitives should return zero. We should discuss it at some point.

Reply:
I'm thinking about doing that in the next PR for all accounting types